dagutils/traversal.go

```diff
@@ -0,0 +1,507 @@
+// what it the current nodes but which nodes (children) have already
+// been visited (assuming always a left to right, zero to `len(Links())`,
+// visiting order).
+//
+// TODO: Add a logic similar to `precalcNextBuf` to request several child
+// nodes in advanced as node promises.
+//
+// TODO: Revisit the name. `Traverser`? (is "traverser" the correct noun?),
+// `Iterator` (this would put to much emphasis on iterate whereas other
+// traversal operations like search are supported), `Topology`?
+//
+// TODO: Encapsulate the position (`path`, `level`, `childIndex`)? It's
+// a big part of the structure, and it should only be one of it, so it
+// would just create a proxy for every move call, e.g., `dt.pos.down()`.
+//
+// TODO: Consider adding methods that would retrieve an attribute indexed
+// by the current `level` to avoid including much indexing like
+// `dt.path[dt.level]` in the code. Maybe another more strong refactoring
+// that would allow to think of the current node in the `path` without
+// concerning at what `level` it's in.
+type Traversal struct {
+
+	// Each member of the slice is the parent of the following member, from
+	// the root to the current node, up until the position marked by `level`. The
+	// slice may contain more elements past that point but they should be
+	// ignored, every time the `level` increases the corresponding child node
+	// is inserted (the slice is not truncated to leverage the already allocated
+	// space). It should *always* have a length bigger than zero with the root
+	// of the DAG at the first position (empty DAGs are not valid).
+	path []ipld.Node
+
+	// This slice has the index of the child each parent in `path` is pointing
+	// to. Its current valid length is determined by `level`. The index in the
+	// parent can be set past all of its child nodes (having a value equal to
+	// `len(Links())`) to signal it has visited (or skipped) all of them. A
+	// leaf node with no children and its index in zero would also comply with
+	// this format.
+	childIndex []uint
+	// TODO: Revisit name, `childPosition`? Do not use the *link* term, put
+	// emphasis in the child (the *what*, not the *how* we get it).
+
+	// Depth of the node of the current position. It grows downwards, root is
+	// level 0, its child level 1, and so on. It controls the effective length
+	// of `path` and `childIndex`.
+	//
+	// A level of -1 signals the start case of a new `Traversal` that hasn't
+	// moved yet (or its position has been reset, see `ResetPosition`).
+	// Although this state is an invalid index to the slices, it allows to
+	// centralize all the visit calls in the `down` move (starting at zero
+	// would require a special visit case inside every traversal operation like
+	// `Iterate()` and `Search`). This value should never be returned to after
+	// the first `down` movement, moving up from the root should always return
+	// `ErrUpOnRoot`.
+	level int
```

Comment:
Why not call this `depth`? It seems like that would be much more straightforward, especially since your comments define it in terms of “depth” ;)

Reply:
Yes, you're right. I avoided the term "depth" at first because I considered it an absolute property, "the depth of the entire DAG", and didn't want to give the sense that this property was being modified here. I could use the name `currentDepth` or `positionDepth`, or yes, just `depth` alone and indicate this is related more to the current position in the DAG than the DAG itself.